### PR TITLE
Allow continuous vars in diff abund

### LINKED ISSUE #41: Differential abundance: allow continuous variables
If the variable is continuous, the group args must be translatable into bins. Group the points by bins

### PR DESCRIPTION
**Pull request by asizemore**
Resolves #41 

Requires three new functions in veupathUtils: findBinStart, findBinEnd, and whichInBin (PR [here](https://github.com/VEuPathDB/veupathUtils/pull/18))

Also resolves some frustrating behavior arising from updating by reference.

### CODE REVIEW

File: R/method-differentialAbundance.R [hunk 1]
```diff
@@ -4,8 +4,8 @@
 #' 
 #' @param data AbsoluteAbundanceData object
 #' @param comparisonVariable string identifying the metadata column to be used when grouping samples.
-#' @param groupA array of strings, each string indicating a value in the comparisonVariable column. Samples with these metadata values will be included in Group A for the differential abundance calculation. Must have no overlap with groupB
-#' @param groupB array of strings, each string indicating a value in the comparisonVariable column. Samples with these metadata values will be included in Group B for the differential abundance calculation. Must have no overlap with groupA.
+#' @param groupA array of strings or a BinList, with each string indicating a value in the comparisonVariable column or each bin a different range for a continuous variable. Samples with these metadata values will be included in Group A for the differential abundance calculation. Must have no overlap with groupB
```
Comment (d-callan) [inline R/method-differentialAbundance.R:7]: why a string or BinList ?? everything besides the bin label is optional iirc, so you can use it for the categorical vars as well. we do that elsewhere for the map..

File: R/method-differentialAbundance.R [hunk 2]
```diff
@@ -53,34 +53,53 @@ setMethod("differentialAbundance", signature("AbsoluteAbundanceData"), function(
       veupathUtils::logWithTime("Replaced NAs with 0", verbose)
     }
 
+    # Err if the class of groupA does not match that of groupB
+    if (!identical(class(groupA), class(groupB))) {
+      stop("groupA and groupB must either both be character vectors or both be BinLists.")
+    }
+
+    # Determine if comparison variable is numeric.
+    isNumericComparisonVar <- identical(class(groupA)[1], "BinList")
```
Comment (d-callan) [inline R/method-differentialAbundance.R]: the heuristics make me uncomfortable. this is the kind of thing that becomes a bug later..

File: R/method-differentialAbundance.R [hunk 3]
```diff
@@ -53,34 +53,53 @@ setMethod("differentialAbundance", signature("AbsoluteAbundanceData"), function(
...
-        groupB <- groupB[groupB %in% uniqueComparisonVariableValues]
-      }
 
-      # Do not allow duplicated values
-      if (!!length(intersect(groupA, groupB))) {
-        veupathUtils::logWithTime("groupA and groupB cannot share members.", verbose)
-        stop()
+      if (!isNumericComparisonVar) {
+
+        # Do not allow duplicated values
```
Comment (d-callan) [inline R/method-differentialAbundance.R:74]: a lot of this feels like validation that could/ should belong to a `Comparator` class

**Review by d-callan — COMMENTED** [review]
overall getting better but i still dont think its quite there. its not really fully leveraging the s4 class system imo. im going to paste this from slack:

imagine a `Comparator` class w slots for `variable` w type `VariableMetadata` , `groupA` w type `BinList` and `groupB` w type `BinList`. bc it has VariableMetadata it can do things like have a method to tell you explicitly if its data shape is continuous or not, which lets you handle dates etc well too. it can have methods like getGroupA which can build a string if you like using either the label or bin start and end as appropriate. that means you can show friendly labels in your ui to select the bins, but guarantee accuracy w the start and ends in the request. it also simplifies your diff abund function call and keeps several related args functionally associated.

**Comment by asizemore** [comment]
Comparator class added. @d-callan i saw your comment about moving this to veupathUtils someday. Do you have an idea for when/where this class will be useful outside of diff abund? There's nothing i know about coming up but maybe there's something in the map you had in mind?

File: R/class-Comparator.R [hunk 4]
```diff
@@ -0,0 +1,78 @@
+
+check_comparator <- function(object) {
```
Comment (d-callan) [inline R/class-Comparator.R:2]: the comparator class could probably go in veupathutils, but im happy to leave it here until we def need it somewhere else.

**Review by d-callan — APPROVED** [review]

File: R/class-Comparator.R (hunk 4 above)
Comment (d-callan) [reply R/class-Comparator.R:2]: nothing specific, just noticing this isnt really a mbio specific class

**Pull request merged**